eth_worker/sql_persistence/interface.py

```diff
@@ -423,16 +445,12 @@ def set_wallet_last_topup_task_uuid(self, wallet_address, task_uuid):
 
         self.session.commit()
 
-    def __init__(self, w3, red, session_factory, PENDING_TRANSACTION_EXPIRY_SECONDS=30):
-
-        self.w3 = w3
+    def __init__(self, red, session, first_block_hash, PENDING_TRANSACTION_EXPIRY_SECONDS=30):
 
         self.red = red
 
-        self.session_factory = session_factory
-
-        self.session = scoped_session(session_factory)
```

Comment:
Any reason for moving the `scoped_session` up a level? (Not a critique, this is fine. Just curious if/how having it this way breaks tests)

Reply:
oh yeah, by shifting injecting the session rather than the factory, it gives u cleaner control over the session lifecycle, as u can now tear it down without having to go digging inside of the interface. Specifically, it allows u to do this: 
```
@pytest.fixture(scope='function')
def db_session():
    # Create the database and the database table

    Base.metadata.create_all(engine)

    session = scoped_session(session_factory)

    yield session

    session.close()
    Base.metadata.drop_all(bind=engine)


@pytest.fixture(scope='function')
def persistence_int(db_session):
    red = redis.Redis.from_url(config.REDIS_URL)

    return SQLPersistenceInterface(
        red=red, session=db_session, first_block_hash='deadbeef01'
    )
```

Reply:
makes unit testing waaay easier